File: Day5/longest_substring.cpp
#include <iostream>
#include <string>
#include <unordered_set>
#include <algorithm>
using namespace std;

// Day 5 — Longest Substring Without Repeating Characters
// Approach: Sliding window with hash set to track unique characters.
// Time: O(n), Space: O(min(m,n)) where m is character set size.
int lengthOfLongestSubstring(const string& s) {
    unordered_set<char> charSet;
    int left = 0;
    int maxLength = 0;
    
    for (int right = 0; right < (int)s.length(); ++right) {
        // Shrink window from left until no duplicates
        while (charSet.find(s[right]) != charSet.end()) {
            charSet.erase(s[left]);
            ++left;
        }
        
        // Add current character and update max length
        charSet.insert(s[right]);
        maxLength = max(maxLength, right - left + 1);
    }
    
    return maxLength;
}

int main() {
    // Test cases from examples
    cout << "Testing Day 5 solutions:\n";
    cout << "abcabcbb -> " << lengthOfLongestSubstring("abcabcbb") << " (expected: 3)\n";
    cout << "bbbbb -> " << lengthOfLongestSubstring("bbbbb") << " (expected: 1)\n";
    cout << "pwwkew -> " << lengthOfLongestSubstring("pwwkew") << " (expected: 3)\n";
    
    // Basic validation
    if (lengthOfLongestSubstring("abcabcbb") != 3 ||
        lengthOfLongestSubstring("bbbbb") != 1 ||
        lengthOfLongestSubstring("pwwkew") != 3 ||
        lengthOfLongestSubstring("") != 0 ||
        lengthOfLongestSubstring("a") != 1) {
        cerr << "Test failed!" << endl;
        return 1;
    }
    
    cout << "All tests passed!\n";
    return 0;
}

File: Day5/longest_substring.py
"""
Day 5 — Longest Substring Without Repeating Characters

Problem: Find the length of the longest substring without repeating characters.
Approach: Sliding window with hash set to track unique characters.
Time: O(n), Space: O(min(m,n)) where m is character set size.
"""


def length_of_longest_substring(s: str) -> int:
    """
    Find length of longest substring without repeating characters.
    
    Args:
        s: Input string
        
    Returns:
        Length of longest substring without repeating characters
    """
    char_set = set()
    left = 0
    max_length = 0
    
    for right in range(len(s)):
        # Shrink window from left until no duplicates
        while s[right] in char_set:
            char_set.remove(s[left])
            left += 1
        
        # Add current character and update max length
        char_set.add(s[right])
        max_length = max(max_length, right - left + 1)
    
    return max_length


if __name__ == "__main__":
    # Test cases from examples
    assert length_of_longest_substring("abcabcbb") == 3  # "abc"
    assert length_of_longest_substring("bbbbb") == 1     # "b"
    assert length_of_longest_substring("pwwkew") == 3    # "wke"
    
    # Edge cases
    assert length_of_longest_substring("") == 0          # empty string
    assert length_of_longest_substring("a") == 1         # single char
    assert length_of_longest_substring("au") == 2        # no repeats
    assert length_of_longest_substring("dvdf") == 3      # "vdf"
    
    print("Day5 longest_substring.py: all tests passed")